src/Columns/ColumnSparse.cpp

```diff
@@ -0,0 +1,762 @@
+#include <Columns/ColumnSparse.h>
+#include <Columns/ColumnsCommon.h>
+#include <Columns/ColumnCompressed.h>
+#include <Columns/ColumnTuple.h>
+#include <Common/WeakHash.h>
+#include <Common/SipHash.h>
+#include <Common/HashTable/Hash.h>
+#include <DataStreams/ColumnGathererStream.h>
+
+#include <algorithm>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int SIZES_OF_COLUMNS_DOESNT_MATCH;
+}
+
+ColumnSparse::ColumnSparse(MutableColumnPtr && values_)
+    : values(std::move(values_)), _size(0)
+{
+    if (!values->empty())
+        throw Exception("Not empty values passed to ColumnSparse, but no offsets passed", ErrorCodes::LOGICAL_ERROR);
+
+    values->insertDefault();
+    offsets = ColumnUInt64::create();
+}
+
+ColumnSparse::ColumnSparse(MutableColumnPtr && values_, MutableColumnPtr && offsets_, size_t size_)
+    : values(std::move(values_)), offsets(std::move(offsets_)), _size(size_)
+{
+    const ColumnUInt64 * offsets_concrete = typeid_cast<const ColumnUInt64 *>(offsets.get());
+
+    if (!offsets_concrete)
+        throw Exception("offsets_column must be a ColumnUInt64", ErrorCodes::LOGICAL_ERROR);
+
+    /// 'values' should contain one extra element: default value at 0 position.
+    if (offsets->size() + 1 != values->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Values size ({}) is inconsistent with offsets size ({})", values->size(), offsets->size());
+
+    if (_size < offsets->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Size of sparse column ({}) cannot be lower than number of non-default values ({})", _size, offsets->size());
+
+#ifndef NDEBUG
+    const auto & offsets_data = getOffsetsData();
+    const auto * it = std::adjacent_find(offsets_data.begin(), offsets_data.end(), std::greater_equal<UInt64>());
+    if (it != offsets_data.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Offsets of ColumnSparse must be strictly sorted");
+#endif
+}
+
+MutableColumnPtr ColumnSparse::cloneResized(size_t new_size) const
+{
+    if (new_size == 0)
+        return ColumnSparse::create(values->cloneEmpty());
+
+    if (new_size >= _size)
+        return ColumnSparse::create(IColumn::mutate(values), IColumn::mutate(offsets), new_size);
+
+    auto res = ColumnSparse::create(values->cloneEmpty());
+    res->insertRangeFrom(*this, 0, new_size);
```

Comment:
Ma be done more optimal.
bin search + 2 * insertRangeFrom

Reply:
I didn't understand what you mean.